cpp/src/parquet/encoding.cc

```diff
@@ -340,32 +340,41 @@ class PlainEncoder<BooleanType> : public EncoderImpl, virtual public BooleanEnco
       throw ParquetException("direct put to boolean from " + values.type()->ToString() +
                              " not supported");
     }
-
     const auto& data = checked_cast<const ::arrow::BooleanArray&>(values);
+
     if (data.null_count() == 0) {
-      PARQUET_THROW_NOT_OK(sink_.Reserve(bit_util::BytesForBits(data.length())));
-      // no nulls, just dump the data
-      ::arrow::internal::CopyBitmap(data.data()->GetValues<uint8_t>(1), data.offset(),
```

Comment:
If the problem here is the incorrect use of `sink_.Reserve` and `sink_.UnsafeAdvance`, why not directly fix them? The current change gives me an illusion that `CopyBitmap` and `FirstTimeBitmapWriter` are the root cause of the bug.

Reply:
You're right...Let me just revert that and correct the `sink_` here.